x/outline-connectivity-app/shared_backend/main.go

```diff
@@ -0,0 +1,292 @@
+	if err == nil {
+		return nil
+	}
+	var record = new(ConnectivityTestError)
+	var testErr *connectivity.TestError
+	if errors.As(err, &testErr) {
+		record.Op = testErr.Op
+		record.PosixError = testErr.PosixError
+		record.Msg = unwrapAll(testErr).Error()
+	} else {
+		record.Msg = err.Error()
+	}
+	return record
+}
+
+func unwrapAll(err error) error {
+	for {
+		unwrapped := errors.Unwrap(err)
+		if unwrapped == nil {
+			return err
+		}
+		err = unwrapped
+	}
+}
+
+func (p Prefix) String() string {
+	runes := make([]rune, len(p))
+	for i, b := range p {
+		runes[i] = rune(b)
+	}
+	return string(runes)
+}
+
+func parseAccessKey(accessKey string) (*sessionConfig, error) {
+	var config sessionConfig
+	accessKeyURL, err := url.Parse(accessKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse access key: %w", err)
+	}
+	var portString string
+	// Host is a <host>:<port> string
+	config.Hostname, portString, err = net.SplitHostPort(accessKeyURL.Host)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse endpoint address: %w", err)
+	}
+	config.Port, err = strconv.Atoi(portString)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse port number: %w", err)
+	}
+	cipherInfoBytes, err := base64.URLEncoding.WithPadding(base64.NoPadding).DecodeString(accessKeyURL.User.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode cipher info [%v]: %v", accessKeyURL.User.String(), err)
+	}
+	cipherName, secret, found := strings.Cut(string(cipherInfoBytes), ":")
+	if !found {
+		return nil, fmt.Errorf("invalid cipher info: no ':' separator")
+	}
+	config.CryptoKey, err = shadowsocks.NewEncryptionKey(cipherName, secret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cipher: %w", err)
+	}
+	prefixStr := accessKeyURL.Query().Get("prefix")
+	if len(prefixStr) > 0 {
+		config.Prefix, err = ParseStringPrefix(prefixStr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse prefix: %w", err)
+		}
+	}
+	return &config, nil
+}
+
+func ParseStringPrefix(utf8Str string) (Prefix, error) {
+	runes := []rune(utf8Str)
+	rawBytes := make([]byte, len(runes))
+	for i, r := range runes {
+		if (r & 0xFF) != r {
+			return nil, fmt.Errorf("character out of range: %d", r)
+		}
+		rawBytes[i] = byte(r)
+	}
+	return rawBytes, nil
+}
+
+/* INFRASTRUCTURE (to be generalized via reflection/generics and moved) */
+type CallInputMessage struct {
+	Method string `json:"method"`
+	Input  string `json:"input"`
+}
+
+type CallOutputMessage struct {
+	Result string   `json:"result"`
+	Errors []string `json:"errors"`
+}
+
+func SendRawCall(rawInputMessage []byte) []byte {
+	var inputMessage CallInputMessage
+
+	parseInputError := json.Unmarshal(rawInputMessage, &inputMessage)
+
+	outputMessage := CallOutputMessage{Result: "", Errors: []string{}}
+
+	if parseInputError != nil {
+		outputMessage.Errors = append(outputMessage.Errors, "SendRawCall: error parsing raw input string")
+	}
+
+	if inputMessage.Method != "ConnectivityTest" {
```

Comment:
Can we do the other way around?

I'd like to see what code is specific to `ConnectivityTest`, so I know how to add a new method.

Reply:
Sorry, I wasn't clear. I meant I'd like to see something like this:
```
if inputMessage.Method == "ConnectivityTest" {
  return HandleConnectivityTest(request, response)
}
```

So it allows the addition of new methods.
Ideally in a way that we don't need to repeat the boilerplate for every method.
Perhaps you have something like:
 
```go
if inputMessage.Method == "ConnectivityTest" {
  return Call(HandleConnectivityTest, request, response)
}
``` 

You can use Generics like:
```go
fun Call[Request, Response](call func(Request, Response) error, request []bytes, response []bytes) error {
  // Unmarshall, call, Marshall
}
```